@miguelgfierro @SimonYansenZhao Previously, I was advised to include `include_groups=False`, but the CI failed on Python 3.8 due to compatibility issues. I’ve opened this PR to address that. Could you please take a look?

```sh
=================================== FAILURES ===================================
________________________ test_negative_feedback_sampler ________________________

self = <pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f14f80d9700>
func = <function negative_feedback_sampler.<locals>.sample_items at 0x7f14f802a670>
args = (), kwargs = {'include_groups': False}
f = <function negative_feedback_sampler.<locals>.sample_items at 0x7f14f80c9af0>

    @Appender(
        _apply_docs["template"].format(
            input="dataframe", examples=_apply_docs["dataframe_examples"]
        )
    )
    def apply(self, func, *args, **kwargs) -> NDFrameT:
        func = com.is_builtin_func(func)
    
        if isinstance(func, str):
            if hasattr(self, func):
                res = getattr(self, func)
                if callable(res):
                    return res(*args, **kwargs)
                elif args or kwargs:
                    raise ValueError(f"Cannot pass arguments to property {func}")
                return res
    
            else:
                raise TypeError(f"apply func should be callable, not '{func}'")
    
        elif args or kwargs:
            if callable(func):
    
                @wraps(func)
                def f(g):
                    with np.errstate(all="ignore"):
                        return func(g, *args, **kwargs)
    
            else:
                raise ValueError(
                    "func must be a callable if args or kwargs are supplied"
                )
        else:
            f = func
    
        # ignore SettingWithCopy here in case the user mutates
        with option_context("mode.chained_assignment", None):
            try:
>               result = self._python_apply_general(f, self._selected_obj)

../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1353: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1402: in _python_apply_general
    values, mutated = self.grouper.apply(f, data, self.axis)
../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/ops.py:767: in apply
    res = f(group)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

g =    userID  itemID  rating
0       1       1       1

    @wraps(func)
    def f(g):
        with np.errstate(all="ignore"):
>           return func(g, *args, **kwargs)
E           TypeError: sample_items() got an unexpected keyword argument 'include_groups'

../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1341: TypeError

During handling of the above exception, another exception occurred:

    def test_negative_feedback_sampler():
        df = pd.DataFrame(
            data={"userID": [1, 2, 3], "itemID": [1, 2, 3], "rating": [5, 5, 5]}
        )
    
        # Test ratio < 1
>       sample_df = negative_feedback_sampler(
            df,
            col_user="userID",
            col_item="itemID",
            col_label="rating",
            ratio_neg_per_user=0.5,
        )

tests/unit/recommenders/datasets/test_pandas_df_utils.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
recommenders/datasets/pandas_df_utils.py:362: in negative_feedback_sampler
    res_df.groupby(col_user)
../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1363: in apply
    return self._python_apply_general(f, self._obj_with_exclusions)
../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1402: in _python_apply_general
    values, mutated = self.grouper.apply(f, data, self.axis)
../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/ops.py:767: in apply
    res = f(group)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

g =    itemID  rating
0       1       1

    @wraps(func)
    def f(g):
        with np.errstate(all="ignore"):
>           return func(g, *args, **kwargs)
E           TypeError: sample_items() got an unexpected keyword argument 'include_groups'

../.venvs/Recommenders/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1341: TypeError
=============================== warnings summary ===============================
../.venvs/Recommenders/lib/python3.8/site-packages/jupyter_client/connect.py:22
  /root/.venvs/Recommenders/lib/python3.8/site-packages/jupyter_client/connect.py:22: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write

../.venvs/Recommenders/lib/python3.8/site-packages/cornac/data/text.py:134
  /root/.venvs/Recommenders/lib/python3.8/site-packages/cornac/data/text.py:134: DeprecationWarning: invalid escape sequence \]
    """

tests/unit/examples/test_notebooks_pyspark.py:17
  /root/Recommenders/tests/unit/examples/test_notebooks_pyspark.py:17: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=5, reruns_delay=2)

tests/unit/examples/test_notebooks_pyspark.py:46
  /root/Recommenders/tests/unit/examples/test_notebooks_pyspark.py:46: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=5, reruns_delay=3)

tests/unit/examples/test_notebooks_pyspark.py:68
  /root/Recommenders/tests/unit/examples/test_notebooks_pyspark.py:68: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=5, reruns_delay=3)

tests/unit/examples/test_notebooks_pyspark.py:80
  /root/Recommenders/tests/unit/examples/test_notebooks_pyspark.py:80: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=5, reruns_delay=2)

tests/unit/examples/test_notebooks_pyspark.py:104
  /root/Recommenders/tests/unit/examples/test_notebooks_pyspark.py:104: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=5, reruns_delay=2)

tests/unit/recommenders/models/test_tfidf_utils.py::test_fit
  /root/Recommenders/tests/unit/recommenders/models/test_tfidf_utils.py:97: DeprecationWarning: Please use `csr_matrix` from the `scipy.sparse` namespace, the `scipy.sparse.csr` namespace is deprecated.
    assert type(model.tfidf_matrix) == scipy.sparse.csr.csr_matrix

tests/unit/recommenders/datasets/test_spark_splitter.py: 8 warnings
tests/unit/recommenders/evaluation/test_spark_evaluation.py: 160 warnings
  /root/.venvs/Recommenders/lib/python3.8/site-packages/pyspark/sql/pandas/utils.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(pandas.__version__) < LooseVersion(minimum_pandas_version):

tests/unit/recommenders/evaluation/test_spark_evaluation.py: 14 warnings
  /root/.venvs/Recommenders/lib/python3.8/site-packages/pyspark/sql/context.py:158: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================== slowest durations ===============================
45.31s call     tests/unit/examples/test_notebooks_pyspark.py::test_evaluation_diversity_runs
44.02s call     tests/unit/examples/test_notebooks_pyspark.py::test_mmlspark_lightgbm_criteo_runs
42.05s call     tests/unit/examples/test_notebooks_pyspark.py::test_als_deep_dive_runs
32.19s call     tests/unit/examples/test_notebooks_pyspark.py::test_als_pyspark_runs
31.08s call     tests/unit/examples/test_notebooks_python.py::test_lightgbm
15.20s call     tests/unit/examples/test_notebooks_pyspark.py::test_evaluation_runs
13.22s call     tests/unit/examples/test_notebooks_pyspark.py::test_data_split_runs
12.80s call     tests/unit/examples/test_notebooks_python.py::test_baseline_deep_dive_runs
11.51s call     tests/unit/examples/test_notebooks_python.py::test_cornac_deep_dive_runs
10.32s call     tests/unit/recommenders/datasets/test_download_utils.py::test_maybe_download_wrong_bytes
8.02s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_map_at_k
7.85s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_ndcg_at_k
7.70s call     tests/unit/recommenders/datasets/test_download_utils.py::test_maybe_download_retry
5.79s call     tests/unit/recommenders/datasets/test_spark_splitter.py::test_stratified_splitter
5.00s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_merge_ranking
4.99s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_precision
4.97s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_recall
4.79s call     tests/unit/examples/test_notebooks_python.py::test_sar_single_node_runs
4.70s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_rmse
4.69s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_auc
4.67s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_logloss
4.64s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_merge_rating
4.63s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_rsquared
4.63s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_mae
4.62s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_exp_var
4.12s call     tests/unit/examples/test_notebooks_python.py::test_sar_deep_dive_runs
3.75s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_diversity_item_feature_vector
2.81s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_init_spark
2.58s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[0-101-True-True]
2.02s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_exp_var
2.00s call     tests/unit/recommenders/utils/test_timer.py::test_timer
1.71s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_precision_at_k
1.68s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_serendipity_item_feature_vector
1.65s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_diversity
1.65s call     tests/unit/recommenders/datasets/test_spark_splitter.py::test_chrono_splitter
1.60s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_value_error_fails
1.53s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_item_serendipity_item_feature_vector
1.47s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_serendipity
1.46s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_serendipity
1.45s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_serendipity_item_feature_vector
1.43s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_diversity_item_feature_vector
1.39s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_item_serendipity
1.39s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution
1.39s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_is_jupyter
1.38s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_with_parameters[parameters3-response2-A]
1.38s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_with_parameters[parameters5-response3-12]
1.38s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_with_parameters[parameters0-response1-8]
1.34s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_novelty
1.27s call     tests/unit/examples/test_notebooks_python.py::test_template_runs
1.25s call     tests/responsible_ai/recommenders/datasets/test_movielens_privacy.py::test_movielens_privacy
1.22s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_with_parameters[parameters1-response1-3.5]
1.22s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_with_parameters[parameters4-response2-100k]
1.21s call     tests/unit/recommenders/utils/test_notebook_utils.py::test_notebook_execution_with_parameters[parameters2-response2-True]
1.15s call     tests/responsible_ai/recommenders/datasets/test_criteo_privacy.py::test_criteo_privacy
1.13s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_ndcg_at_k
1.11s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_diversity
1.04s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_item_novelty
1.02s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_auc
1.02s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_diversity_item_feature_vector
1.01s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_logloss
1.01s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_map_at_k
0.96s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_map
0.95s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_rmse
0.91s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_python_match[10-0-None0]
0.91s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_python_match[10-0-None1]
0.86s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_mae
0.85s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_rsquared
0.83s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_recall_at_k
0.80s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_python_match[3-0-None]
0.78s call     tests/unit/recommenders/datasets/test_spark_splitter.py::test_timestamp_splitter
0.72s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_python_match[10-0-3]
0.67s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_get_top_k_items_largek
0.59s call     tests/unit/recommenders/datasets/test_spark_splitter.py::test_random_splitter
0.56s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[3-101-True-True]
0.51s call     tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_get_top_k_items
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_map_at_k
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_merge_rating
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_rmse
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_precision
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_ndcg_at_k
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_rsquared
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_merge_ranking
0.47s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_recall
0.46s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_mae
0.46s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_python_exp_var
0.43s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_distributional_coverage
0.40s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_catalog_coverage
0.39s call     tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_init_spark_rating_eval
0.33s call     tests/unit/recommenders/datasets/test_spark_splitter.py::test_min_rating_filter
0.30s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_load_spark_df_mock_100__with_custom_param__succeed
0.26s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[3-101-True-False]
0.26s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[3-101-False-True]
0.26s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[0-101-True-False]
0.26s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[3-101-False-False]
0.25s call     tests/unit/recommenders/models/test_tfidf_utils.py::test_fit
0.24s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_get_top_k_items_largek
0.24s setup    tests/performance/recommenders/evaluation/test_python_evaluation_time_performance.py::test_get_top_k_items
0.22s teardown tests/unit/examples/test_notebooks_pyspark.py::test_evaluation_diversity_runs
0.21s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[0-101-False-False]
0.20s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_load_spark_df_mock_100__with_default_param__succeed
0.20s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__data_serialization_default_param
0.20s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[0-101-False-True]
0.20s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[10-101-True-True]
0.20s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[10-101-True-False]
0.19s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[10-101-False-False]
0.18s call     tests/data_validation/recommenders/datasets/test_wikidata.py::test_search_wikidata_incorrect
0.17s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__return_success[10-101-False-True]
0.14s call     tests/data_validation/recommenders/datasets/test_movielens.py::test_mock_movielens_schema__get_spark_df__store_tmp_file
0.12s call     tests/unit/recommenders/models/test_cornac_utils.py::test_recommend_k_items
0.12s setup    tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-cooccurrence-count]
0.10s call     tests/unit/recommenders/tuning/test_nni_utils.py::test_check_experiment_status_running
0.10s call     tests/unit/recommenders/tuning/test_nni_utils.py::test_check_stopped_timeout
0.10s call     tests/unit/recommenders/tuning/test_nni_utils.py::test_check_experiment_status_no_more_trial
0.10s call     tests/unit/recommenders/tuning/test_nni_utils.py::test_check_metrics_written_timeout
0.08s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_diversity_item_feature_vector
0.07s setup    tests/unit/recommenders/datasets/test_spark_splitter.py::test_stratified_splitter
0.07s call     tests/unit/recommenders/models/test_tfidf_utils.py::test_tokenize_text
0.07s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_user_affinity
0.07s setup    tests/unit/recommenders/models/test_tfidf_utils.py::test_get_tokens
0.06s call     tests/data_validation/recommenders/datasets/test_wikidata.py::test_query_entity_links
0.06s call     tests/unit/recommenders/models/test_cornac_utils.py::test_predict
0.06s setup    tests/unit/recommenders/datasets/test_spark_splitter.py::test_min_rating_filter
0.05s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_item_serendipity
0.05s call     tests/unit/recommenders/utils/test_plot.py::test_line_graph
0.05s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_ndcg_at_k
0.05s setup    tests/unit/recommenders/datasets/test_spark_splitter.py::test_timestamp_splitter
0.05s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_serendipity
0.05s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_item_novelty
0.05s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_diversity
0.05s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_diversity
0.05s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-cosine-cos]
0.05s setup    tests/unit/recommenders/datasets/test_spark_splitter.py::test_random_splitter
0.04s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_serendipity
0.04s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_serendipity_item_feature_vector
0.04s call     tests/unit/recommenders/models/test_tfidf_utils.py::test_get_top_k_recommendations
0.04s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_novelty
0.04s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_serendipity_item_feature_vector
0.04s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_user_item_serendipity_item_feature_vector
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-mutual information-mi]
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-inclusion index-incl]
0.04s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_precision_at_k
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-inclusion index-incl]
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-jaccard-jac]
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-mutual information-mi]
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-cosine-cos]
0.04s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-lexicographers mutual information-lex]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-lift-lift]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-cooccurrence-count]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-lexicographers mutual information-lex]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-mutual information-mi]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-lift-lift]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-cosine-cos]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-lexicographers mutual information-lex]
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_distributional_coverage
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-lift-lift]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_recommend_k_items[3-jaccard-jac]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-inclusion index-incl]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[1-cooccurrence-count]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-jaccard-jac]
0.03s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_sar_item_similarity[3-cooccurrence-count]
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_catalog_coverage
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_recall_at_k
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_exp_var
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_map_at_k
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_map
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_init_spark_rating_eval
0.03s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_precision_at_k
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_map_at_k
0.02s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_rmse
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_map
0.02s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_rsquared
0.02s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_mae
0.02s setup    tests/unit/recommenders/evaluation/test_spark_evaluation.py::test_spark_ndcg_at_k
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_serendipity_item_feature_vector
0.02s call     tests/unit/recommenders/datasets/test_python_splitter.py::test_chrono_splitter
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_user_serendipity_item_feature_vector
0.02s call     tests/unit/recommenders/datasets/test_download_utils.py::test_maybe_download_maybe
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_recall_at_k
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_user_diversity
0.02s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_user_item_serendipity_item_feature_vector
0.02s call     tests/unit/recommenders/datasets/test_download_utils.py::test_maybe_download
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_user_diversity_item_feature_vector
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_novelty
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_user_item_serendipity
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_diversity_item_feature_vector
0.01s call     tests/unit/recommenders/datasets/test_python_splitter.py::test_stratified_splitter
0.01s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_get_item_based_topk
0.01s setup    tests/unit/recommenders/datasets/test_sparse.py::test_df_to_sparse
0.01s setup    tests/unit/recommenders/datasets/test_python_splitter.py::test_random_splitter
0.01s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_get_normalized_scores
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_serendipity
0.01s call     tests/unit/recommenders/datasets/test_sparse.py::test_sparse_to_df
0.01s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_get_topk_most_similar_users
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_merge_ranking
0.01s call     tests/unit/recommenders/models/test_vowpal_wabbit.py::test_fit_and_predict
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_user_serendipity
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_item_novelty
0.01s call     tests/unit/recommenders/datasets/test_pandas_df_utils.py::test_csv_to_libffm
0.01s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_predict[lift-True]
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_auc
0.01s call     tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_logloss
0.01s call     tests/unit/recommenders/models/test_sar_singlenode.py::test_fit[lift-True]

(597 durations < 0.005s hidden.  Use -vv to show these durations.)
=========================== short test summary info ============================
FAILED tests/unit/recommenders/datasets/test_pandas_df_utils.py::test_negative_feedback_sampler
```

--- a/recommenders/datasets/pandas_df_utils.py
+++ b/recommenders/datasets/pandas_df_utils.py
@@ -331,10 +331,6 @@ def negative_feedback_sampler(
     rng = np.random.default_rng(seed=seed)
 
     def sample_items(user_df):
-        # Save group key before dropping the groupby column (drop returns a new DataFrame
-        # without the .name attribute set by GroupBy)
-        user_id = user_df.name
-        user_df = user_df.drop(columns=[col_user])
         # Sample negative items for the data frame restricted to a specific user
         n_u = len(user_df)
         neg_sample_size = (
@@ -350,13 +346,13 @@ def negative_feedback_sampler(
         new_items = np.setdiff1d(items_sample, user_df[col_item])[:neg_sample_size]
         new_df = pd.DataFrame(
             data={
-                col_user: user_id,
+                col_user: user_df.name,
                 col_item: new_items,
                 col_label: neg_value,
             }
         )
         combined = pd.concat(
-            [user_df.assign(**{col_user: user_id}), new_df], ignore_index=True
+            [user_df.assign(**{col_user: user_df.name}), new_df], ignore_index=True
         )
         return combined[[col_user, col_item, col_label]]
 
@@ -364,7 +360,7 @@ def negative_feedback_sampler(
     res_df[col_label] = pos_value
 
     return (
-        res_df.groupby(col_user)
+        res_df.groupby(col_user)[[col_item, col_label]]
         .apply(sample_items)
         .reset_index(drop=True)
         .rename(columns={col_label: col_feedback})

--- a/recommenders/datasets/pandas_df_utils.py
+++ b/recommenders/datasets/pandas_df_utils.py
@@ -331,6 +331,10 @@ def negative_feedback_sampler(
     rng = np.random.default_rng(seed=seed)
 
     def sample_items(user_df):
+        # Save group key before dropping the groupby column (drop returns a new DataFrame
+        # without the .name attribute set by GroupBy)
+        user_id = user_df.name
+        user_df = user_df.drop(columns=[col_user])
         # Sample negative items for the data frame restricted to a specific user
         n_u = len(user_df)
         neg_sample_size = (
@@ -346,23 +350,22 @@ def negative_feedback_sampler(
         new_items = np.setdiff1d(items_sample, user_df[col_item])[:neg_sample_size]
         new_df = pd.DataFrame(
             data={
-                col_user: user_df.name,
+                col_user: user_id,
                 col_item: new_items,
                 col_label: neg_value,
             }
         )
         combined = pd.concat(
-            [user_df.assign(**{col_user: user_df.name}), new_df], ignore_index=True
+            [user_df.assign(**{col_user: user_id}), new_df], ignore_index=True
         )
         return combined[[col_user, col_item, col_label]]
 
     res_df = df.copy()
     res_df[col_label] = pos_value
-    apply_kwargs = {"include_groups": False} if pd.__version__ >= "2.2.0" else {}
 
     return (
         res_df.groupby(col_user)
-        .apply(sample_items, **apply_kwargs)
+        .apply(sample_items)
         .reset_index(drop=True)
         .rename(columns={col_label: col_feedback})
     )

